packages/@react-spectrum/menu/src/useSafelyMouseToSubmenu.ts

```diff
@@ -0,0 +1,107 @@
+import {RefObject, useEffect, useRef} from 'react';
+import {useResizeObserver} from '@react-aria/utils';
+
+interface SafelyMouseToSubmenuOptions {
+  /** Ref for the parent menu. */
+  menuRef: RefObject<Element>,
+  /** Ref for the submenu. */
+  submenuRef: RefObject<Element>,
+  /** Whether the submenu is open. */
+  isOpen: boolean
+}
+
+/**
+ * Allows the user to move their pointer to the submenu without it closing when their mouse leaves the trigger element.
+ * Prevents pointer events from going to the underlying menu if the user is moving their pointer towards the sub-menu.
+ */
+export function useSafelyMouseToSubmenu(options: SafelyMouseToSubmenuOptions) {
+  let {menuRef, submenuRef, isOpen} = options;
+  let prevPointerPos = useRef<{x: number, y: number, time: number}>(null);
+  let triggerRect = useRef<DOMRect>(null);
+  let submenuRect = useRef<DOMRect>(null);
+  let lastProcessedTime = useRef<number>(0);
+  let isPointerMovingTowardsSubmenu = useRef<boolean>(false);
+
+  useResizeObserver({
+    ref: submenuRef,
+    onResize: () => submenuRect.current = menuRef.current.getBoundingClientRect()
+  });
+
+  useEffect(() => {
+    let menu = menuRef.current;
+    let submenu = submenuRef.current;
+
+    if (!menu || !submenu || !isOpen) {
+      return;
+    }
+
+    submenuRect.current = submenu.getBoundingClientRect();
+
+    let onPointerMove = (e: PointerEvent) => {
+      let currentTime = Date.now();
+
+      // Throttle
+      if (currentTime - lastProcessedTime.current < 16) {
+        return;
+      }
+
+      // Measure trigger element if we haven't already.
+      if (!triggerRect.current) {
+        triggerRect.current = (e.target as HTMLElement).closest('li').getBoundingClientRect();
```

Comment:
may not be able to rely on `li`, could just be a div with a role

given that we have one of these hooks per trigger, we should be able to just pass in the triggerRef